UPGRADE.md

```diff
@@ -1411,6 +1431,124 @@ const Layout = (props) => {
+
+```diff
+const MyCustomCreate = () => {
+    const createControllerProps = useCreateController();
+    const notify = useNotify();
++    const redirect = useRedirect();
+
+    const handleSubmit = (values) => {
+-        createControllerProps.save(values, 'show', {
++        createControllerProps.save(values, {
+            onSuccess: (data) => {
+                notify('Success');
++                redirect('show', '/posts', data.id);
+            }
+        })
+    }
+
+    return (
+        <CreateContextProvider value={createControllerProps}>
+            <Form
+                onSubmit={handleSubmit}
+                render={formProps => (
+                    <form onSubmit={props.handleSubmit}>
+                        ...
+                    </form>
+                )}
+            />
+        </CreateContextProvider>
+    )
+}
+```
+
+### `<FormContext>`, `<FormContextProvider>` and `useFormContext` Have Been Removed
+
+As the `save` and `saving` properties were already available through the `<SaveContext>` component and its `useSaveContext` hook, we removed the `<FormContext>`, `<FormContextProvider>` components as well the `useFormContext` hook.
+
+The functions around form groups management we extracted into the `<FormGroupsProvider>` component and `useFormGroups` hook.
+
+### The `redirect` prop Has Been Removed From `FormWithRedirect`, `SimpleForm`, `TabbedForm` and `SaveButton`
+
+The `FormWithRedirect`, `SimpleForm`, `TabbedForm` and `SaveButton` don't have a `redirect` prop anymore. You should handle redirection either:
+- in the submit handler if you provided a custom one, leveraging the `useRedirect` hook
+- at the `Create` or `Edit` level as they now accept a `redirect` prop
+- in the `SaveButton` side effects props (`onSuccess` or `onFailure`), leveraging the `useRedirect` hook
+
+If you had the `redirect` prop set on the form component, move it to `Create` or `Edit` component:
+
+```diff
+-<Create>
++<Create redirect="edit">
+-    <SimpleForm redirect="edit">
++    <SimpleForm>
+        <TextInput source="name">
+    </SimpleForm>
+</Create>
+```
+
+If you had the `redirect` prop set on the `SaveButton`, provide a `onSuccess` prop:
+
+```diff
+const PostCreateToolbar = props => {
++    const notify = useNotify();
++    const redirect = useRedirect();
+    return (
+        <Toolbar {...props}>
+            <SaveButton
+                label="post.action.save_and_edit"
+-                redirect="edit"
++                onSuccess={data => {
++                    notify('ra.notification.updated', {
++                        type: 'info',
++                        messageArgs: { smart_count: 1 },
++                        undoable: true,
++                    });
++                    redirect('edit', '/posts', data.id)
++                }}
+            />
+        </Toolbar>
+    );
+};
+```
+
+### The Form Components `save` Prop Has Been Renamed to `onSubmit`:
+
```

Comment:
You should probably add that this prop was only useful if SimpleForml was used in standalone, not as child of Create of Edit. In the general case, there is nothing to do.

Reply:
Added